### ISSUE TITLE
PersistentEntity must not be null! in complete sample

### ISSUE THREAD

**Issue by another-name** [spring-guides/gs-accessing-gemfire-data-rest#1]
I have the error 

```
PersistentEntity must not be null!
```

when trying to play around with this example. I'm new to spring and gemfire so any help would be appreciated.

**Comment by miconner** [comment]
Same error here,  ./gradle bootRun starts ok,  running the first POST example throws an error : 

curl -i -X POST -H "Content-Type:application/json" -d '{  "firstName" : "Frodo",  "lastName" : "Baggins" }' http://localhost:8080/people

HTTP/1.1 500 Internal Server Error
Server: Apache-Coyote/1.1
Content-Type: application/hal+json;charset=UTF-8
Transfer-Encoding: chunked
Date: Tue, 01 Dec 2015 22:47:33 GMT
Connection: close
{"cause":null,"message":"PersistentEntity must not be null!"}

**Comment by gregturn** [comment]
I see it as well. @olivergierke  or @jxblum, can you look at this? I can't make heads or tails of where it's breaking.

**Comment by jxblum** [comment]
Will take a look shortly.

**Comment by jxblum** [comment]
I have reproduced the issue and I am investigating further.  It is clear that it fails because no `PersistentEntity` was registered for the `Person` domain type, however, it is not clear why the domain type was not registered as a `PersistentEntity` based on the `PersonRepository` definition on startup, unless for some strange reason the `PersonRepository` was not resolved (nah).  I need to jog my memory again how (when) a `PersistentEntity` is registered in the core Spring Data infrastructure (doh).

**Comment by jxblum** [comment]
Alright, so there are seemingly a few issues here:

1 First, the `MappingContext` (and in particular, the `GemfireMappingContext`) created by the _Spring Data GemFire_ [GemfireRepositoryFatoryBean](https://github.com/spring-projects/spring-data-gemfire/blob/master/src/main/java/org/springframework/data/gemfire/repository/support/GemfireRepositoryFactoryBean.java#L92-L93) (when a bean of type `GemfireMappingContext` has **not** been explicitly defined in the Spring context config) during the creation of the application's `PersonRepository` instance is not the same `GemfireMappingContext` "resolved" by the `RepositoryRestMvcConfiguration` [persistentEntities](https://github.com/spring-projects/spring-data-rest/blob/master/spring-data-rest-webmvc/src/main/java/org/springframework/data/rest/webmvc/config/RepositoryRestMvcConfiguration.java#L195-L196) bean.

This is mostly due to the fact that the "default" `GemfireMappingContext` created by the `GemfireRepositoryFactoryBean` (when 1 is not explicitly defined) is **not** registered as a bean in the Spring context.  Therefore, the `persistentEntities` bean is unable to resolve the appropriate `MappingContext` to use in order to resolve the `PersistentEntity` for the `Person` REST Resource (entity).

2 No matter; a minor inconvenience.  So, I decided to explicitly define a `(Gemfire)MappingContext` bean in the application's Spring context and configure the SD infrastructure (with the `@EnableGemfireRepositories` [mappingContextRef](https://github.com/spring-projects/spring-data-gemfire/blob/master/src/main/java/org/springframework/data/gemfire/repository/config/EnableGemfireRepositories.java#L137) attribute) to "refer" to this bean definition like so...

```
@EnableGemfireRepositories(mappingContextRef = "gemfireMappingContext")
public class Application {
...
    @Bean
    public GemfireMappingContext gemfireMappingContext() {
        return new GemfireMappingContext();
    }
}
```

While this allowed the [persistentEntities](https://github.com/spring-projects/spring-data-rest/blob/master/spring-data-rest-webmvc/src/main/java/org/springframework/data/rest/webmvc/config/RepositoryRestMvcConfiguration.java#L195-L196) bean to appropriately resolve the explicitly defined `(Gemfire)MappingContext` bean in the Spring context, it was not injected as I expected into the [GemfireRepositoryFactoryBean](https://github.com/spring-projects/spring-data-gemfire/blob/master/src/main/java/org/springframework/data/gemfire/repository/support/GemfireRepositoryFactoryBean.java#L68-L70) before the construction of the application `PersonRepository` instance.

Therefore, again the SD REST/Repository infrastructure and the GemFire-backed `PersonRepository` do not share the same `(Gemfire)MappingContext` that enables the `Person` Resource, `PersistentEntity` to be resolved.

In fact, the `PersonRepository` bean and `persistentEntities` bean now have 2 different `(Gemfire)MappingContext` instances.

3 Even more troubling is that I could seemingly provide any value to the `@EnableGemfireRepositories` [mappingContextRef](https://github.com/spring-projects/spring-data-gemfire/blob/master/src/main/java/org/springframework/data/gemfire/repository/config/EnableGemfireRepositories.java#L137) attribute (e.g. 'nonExistingBean') and the SD repo infrastructure did not take the attribute value into account, i.e. verifying the attribute value is an actual, valid bean reference to an existing `MappingContext` bean in the Spring context.  Mainly, I was curious why the corresponding `GemfireRepositoryFactoryBean` property (`gemfireMappingContext`) was not being injected with the explicitly defined `(Gemfire)MappingContext` bean and wanted to make sure that SD repo infrastructure at least acknowledged I was explicitly setting a bean reference.

Perhaps if anything, the [GemfireRepositoryFactoryBean](https://github.com/spring-projects/spring-data-gemfire/blob/master/src/main/java/org/springframework/data/gemfire/repository/support/GemfireRepositoryFactoryBean.java#L92-L93) should be "registering" the "default" `GemfireMappingContext` instance as a bean in the (application's) Spring context.

More to come...

**Comment by jxblum** [comment]
So, the fix for _#2_, from my previous comment (above), is to have _Spring Data GemFire's_ [GemfireRepositoryExtension](https://github.com/spring-projects/spring-data-gemfire/blob/master/src/main/java/org/springframework/data/gemfire/repository/config/GemfireRepositoryConfigurationExtension.java) override the _Spring Data Commons_ `RepositoryConfigurationExtensionSupport` class's [postProcess(:BeanDefinitionBuilder, :AnnotationRepositoryConfigurationSource)](https://github.com/spring-projects/spring-data-commons/blob/master/src/main/java/org/springframework/data/repository/config/RepositoryConfigurationExtensionSupport.java#L140) no-op method.  In this way, both the [persistentEntitties](https://github.com/spring-projects/spring-data-rest/blob/master/spring-data-rest-webmvc/src/main/java/org/springframework/data/rest/webmvc/config/RepositoryRestMvcConfiguration.java#L195-L196) bean and the `PersonRepository` instance (bean) share the same `(Gemfire)MappingContext` when explicitly defined like so...

```
@EnableGemfireRepositories(mappingContextRef = "gemfireMappingContext")
public class Application {
...
    @Bean
    public GemfireMappingContext gemfireMappingContext() {
        return new GemfireMappingContext();
    }
}
```

However, I am still running into a "_PersistentEntity must not be null_" `IllegalArgumentException` when creating the `RepositoryRestMvcConfiguration` [resourceMappings](https://github.com/spring-projects/spring-data-rest/blob/2.4.1.RELEASE/spring-data-rest-webmvc/src/main/java/org/springframework/data/rest/webmvc/config/RepositoryRestMvcConfiguration.java#L570-L573) bean...

```
Caused by: java.lang.IllegalArgumentException: PersistentEntity must not be null!
    at org.springframework.util.Assert.notNull(Assert.java:115) ~[spring-core-4.2.3.RELEASE.jar:4.2.3.RELEASE]
    at org.springframework.data.rest.core.mapping.RepositoryAwareResourceMetadata.<init>(RepositoryAwareResourceMetadata.java:52) ~[spring-data-rest-core-2.4.1.RELEASE.jar:na]
    at org.springframework.data.rest.core.mapping.RepositoryResourceMappings.populateCache(RepositoryResourceMappings.java:83) ~[spring-data-rest-core-2.4.1.RELEASE.jar:na]
    at org.springframework.data.rest.core.mapping.RepositoryResourceMappings.<init>(RepositoryResourceMappings.java:71) ~[spring-data-rest-core-2.4.1.RELEASE.jar:na]
    at org.springframework.data.rest.core.mapping.RepositoryResourceMappings.<init>(RepositoryResourceMappings.java:53) ~[spring-data-rest-core-2.4.1.RELEASE.jar:na]
    at org.springframework.data.rest.webmvc.config.RepositoryRestMvcConfiguration.resourceMappings(RepositoryRestMvcConfiguration.java:572) ~[spring-data-rest-webmvc-2.4.1.RELEASE.jar:na]
    at org.springframework.boot.autoconfigure.data.rest.SpringBootRepositoryRestMvcConfiguration$$EnhancerBySpringCGLIB$$f1c9a246.CGLIB$resourceMappings$7(<generated>) ~[spring-boot-autoconfigure-1.3.0.RELEASE.jar:1.3.0.RELEASE]
    at org.springframework.boot.autoconfigure.data.rest.SpringBootRepositoryRestMvcConfiguration$$EnhancerBySpringCGLIB$$f1c9a246$$FastClassBySpringCGLIB$$7de6ddf2.invoke(<generated>) ~[spring-boot-autoconfigure-1.3.0.RELEASE.jar:1.3.0.RELEASE]
    at org.springframework.cglib.proxy.MethodProxy.invokeSuper(MethodProxy.java:228) ~[spring-core-4.2.3.RELEASE.jar:4.2.3.RELEASE]
    at org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:318) ~[spring-context-4.2.3.RELEASE.jar:4.2.3.RELEASE]
    at org.springframework.boot.autoconfigure.data.rest.SpringBootRepositoryRestMvcConfiguration$$EnhancerBySpringCGLIB$$f1c9a246.resourceMappings(<generated>) ~[spring-boot-autoconfigure-1.3.0.RELEASE.jar:1.3.0.RELEASE]
    at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_65]
    at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:1.8.0_65]
    at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_65]
    at java.lang.reflect.Method.invoke(Method.java:497) ~[na:1.8.0_65]
    at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:162) ~[spring-beans-4.2.3.RELEASE.jar:4.2.3.RELEASE]
    ... 56 common frames omitted
```

**Comment by jxblum** [comment]
And finally, I found the missing piece to the puzzle (phew!).  Yet another bug in _Spring Data GemFire_ Repository infrastructure implementation logic.  Specifically, the `GemfireRepositoryFactoryBean` [setGemfireMappingContext(..)](https://github.com/spring-projects/spring-data-gemfire/blob/master/src/main/java/org/springframework/data/gemfire/repository/support/GemfireRepositoryFactoryBean.java#L68-L70) fails to the set the `mappingContext` field in the _Spring Data Commons_ `RepositoryFactoryBeanSupport` class, as it does when constructing the "default" `MappingContext` instance [here](https://github.com/spring-projects/spring-data-gemfire/blob/master/src/main/java/org/springframework/data/gemfire/repository/support/GemfireRepositoryFactoryBean.java#L93), which is triggered when a `MappingContext` bean is not explicitly defined in the application's Spring context.  Therefore, when the `PersistentEntity` for the REST Resource is summoned from the Repository meta-data (information) in [getPersistentEntity()](https://github.com/spring-projects/spring-data-commons/blob/master/src/main/java/org/springframework/data/repository/core/support/RepositoryFactoryBeanSupport.java#L178-L185) method, the result is _null_, thus, leading to the Exception above (in my previous comment).

Of course, this is only a problem when you explicitly define a `(Gemfire)MappingContext` bean in the Spring context as previously illustrated.  However, this was necessary since the "default" `(Gemfire)MappingContext` [created](https://github.com/spring-projects/spring-data-gemfire/blob/master/src/main/java/org/springframework/data/gemfire/repository/support/GemfireRepositoryFactoryBean.java#L92-L93) by the `GemfireRepositoryFactoryBean` was not "registered" as a bean in the Spring context (and probably should).

**Comment by jxblum** [comment]
At last, I filed 3 JIRA tickets (bugs) on the _Spring Data GemFire_ to address the various issues described in my previous comments above:

[SGF-448 - GemfireRepositoryConfigurationExtension needs to override the RepositoryConfigurationExtensionSupport postProcess(:BeanDefinitionBuilder, :AnnotationRepositoryConfigurationSource) method.](https://jira.spring.io/browse/SGF-448)

[SGF-449 - GemfireRepositoryFactoryBeanSupport.setGemfireMappingContext needs to call RepositoryFactoryBeanSupport.setMappingContext.](https://jira.spring.io/browse/SGF-449)

[SGF-450 - GemfireRepositoryFactoryBeanSupport needs to explicitly register the "default" GemfireMappingContext when not explicitly defined as a bean in the Spring context.](https://jira.spring.io/browse/SGF-450)

I am not complete certain about the last 1 (SGF-450) yet, but I already have fixes for the other 2.  I will wrap this up tomorrow.  Once the fixes are committed, you will need to tweak the guide dependencies to pick up the SDG BUILD-SNAPSHOT containing the fixes.

I will follow up with those details tomorrow when I wrap up the fixes.

**Comment by jxblum** [comment]
Alright, all 3 _Spring Data GemFire_ issues ([SGF-448](https://jira.spring.io/browse/SGF-448), [SGF-449](https://jira.spring.io/browse/SGF-449) and [SGF-450](https://jira.spring.io/browse/SGF-450)) have been resolved.

For the time being (until Spring Boot 1.3.1 is released, which should possibly [change to the Spring Data-Gosling-SR2](https://github.com/spring-projects/spring-boot/blob/master/spring-boot-dependencies/pom.xml#L135) (service release 2), which [corresponds to Spring Data GemFire 1.7.2](https://github.com/spring-projects/spring-data-commons/wiki/Release-Train-Gosling#participating-modules), the location of the fix), you will need to modify the 'Accessing GemFire Data with REST' Maven POM repo and dependencies as follows...

```
    <properties>
        <java.version>1.8</java.version>
        <spring-data-gemfire.version>1.8.0.BUILD-SNAPSHOT</spring-data-gemfire.version>
    </properties>
...
    <dependencies>
        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-data-rest</artifactId>
        </dependency>
<!--
        <dependency>
            <groupId>org.springframework.boot</groupId>
            <artifactId>spring-boot-starter-data-gemfire</artifactId>
        </dependency>
-->
        <dependency>
            <groupId>org.springframework.data</groupId>
            <artifactId>spring-data-gemfire</artifactId>
            <version>${spring-data-gemfire.version}</version>
        </dependency>
    </dependencies>
...
    <repositories>
         <repository>
             <id>spring-releases</id>
             <url>https://repo.spring.io/libs-release</url>
         </repository>
         <repository>
             <id>spring-snapshot</id>
             <url>https://repo.spring.io/libs-snapshot</url>
         </repository>
     </repositories>
```

I ran the example and all is well using the SDG 1.8.0.BUILD-SNAPSHOT.  Likewise you can use 1.7.2.BUILD-SNAPSHOT.

If there are any more issues, please post back to this thread.

Thanks!

**Comment by jxblum** [comment]
@gregturn I was thinking of leaving this issue open until Spring Boot (e.g. 1.3.1) picks up the new Service Release (i.e. Gosling SR2, which will contain SDG 1.7.2 with the fix).  Unfortunately, Gosling SR1 just went out a few weeks back.  Any thoughts?

**Comment by gregturn** [comment]
That's fine by me.

**Comment by gregturn** [comment]
Thanks for digging out a fix for this.

**Comment by miconner** [comment]
Thank you, it's working for me also with updates to the complete/build.gradle file 

```
repositories {
    mavenCentral()
    maven { url "https://repo.spring.io/libs-release" }
    maven { url "https://repo.spring.io/libs-snapshot" }
}

sourceCompatibility = 1.8
targetCompatibility = 1.8

dependencies {
    compile("org.springframework.boot:spring-boot-starter-data-rest")
    compile("org.springframework.data:spring-data-gemfire:1.8.0.BUILD-SNAPSHOT")
    testCompile("junit:junit")
}
```

I was able to get this example to work connecting to a remote GemFire repo also with some updates to Application.java.   I don't know if this is 100% right but was happy to see how easy this was.

```
    @Bean
    ClientCacheFactoryBean clientCacheFactoryBean() {
        ClientCacheFactoryBean bean = new ClientCacheFactoryBean();

        bean.setPdxPersistent(true);
        bean.setPdxReadSerialized(true);

        try {
            bean.setPool(poolFactoryBean().getObject());
        }
        catch (Exception e){
            e.printStackTrace();
        }

        return bean;
    }
    @Bean
    ClientRegionFactoryBean<String, PdxInstance> clientRegionFactoryBean(final GemFireCache cache) {
        return new ClientRegionFactoryBean<String, PdxInstance>(){{
            setCache(cache);
            setName("hello");
            setClose(false);
            setShortcut(ClientRegionShortcut.PROXY);
        }};
    }

    @Bean
    PoolFactoryBean poolFactoryBean() {
        PoolFactoryBean bean = new PoolFactoryBean();
        ConnectionEndpoint locator = new ConnectionEndpoint(<HOSTNAME>, <PORT>);
        ConnectionEndpointList list = new ConnectionEndpointList(locator);
        bean.setLocatorEndpointList(list);
        return bean;
    }

    @Bean
    JSONRegionAdvice jsonRegionAdvice() {
        JSONRegionAdvice advice = new JSONRegionAdvice();
        advice.setIncludedRegionNames(Arrays.asList("hello"));

        return advice;
    }

```

**Comment by washawu** [comment]
@miconner: "I was able to get this example to work connecting to a remote GemFire repo also with some updates to Application.java. I don't know if this is 100% right but was happy to see how easy this was".

I applied your code changes to the sample project and made it running against a "remote" GemFrire repo running on localhost:40404 which was started in gfsh. The Application started fine and was able to locate the "hello" region.  Unfortunately the dreaded Exception showed up when I ran this: 
       C:>curl http://localhost:8080/people
       {"cause":null,"message":"PersistentEntity must not be null!"}

So the fix jxblum put in works only against a local cache server but still fails when the Application is running as a cache client.

Am I missing something? Please comment, thanks.

[Application.txt](https://github.com/spring-guides/gs-accessing-gemfire-data-rest/files/59753/Application.txt)

**Comment by miconner** [comment]
@washawu,  check the dependencies on build.gradle, should include the new snapshot version and no longer reference spring-boot-starter-data-gemfire.   It worked for me on the remote server.  I also created the region first :  "gfsh>create region --name=hello --type=REPLICATE_PERSISTENT" before running the example, good luck.

**Comment by washawu** [comment]
@miconner, jxblum's fix works fine with local cache server. But it doesn't work with remote cache server the way you suggested. It failed with this Exception when I ran "curl http://localhost:8080/people/1":

```
java.lang.IllegalArgumentException: PersistentEntity must not be null!
at org.springframework.util.Assert.notNull(Assert.java:115) ~[spring-core-4.2.3.RELEASE.jar:4.2.3.RELEASE]
at org.springframework.data.rest.webmvc.PersistentEntityResource$Builder.<init>(PersistentEntityResource.java:139) ~[spring-data-rest-webmvc-2.4.1.RELEASE.jar:na]
at org.springframework.data.rest.webmvc.PersistentEntityResource$Builder.<init>(PersistentEntityResource.java:122) ~[spring-data-rest-webmvc-2.4.1.RELEASE.jar:na]
at org.springframework.data.rest.webmvc.PersistentEntityResource.build(PersistentEntityResource.java:114) ~[spring-data-rest-webmvc-2.4.1.RELEASE.jar:na]
at org.springframework.data.rest.webmvc.PersistentEntityResourceAssembler.wrap(PersistentEntityResourceAssembler.java:102) ~[spring-data-rest-webmvc-2.4.1.RELEASE.jar:na]
at org.springframework.data.rest.webmvc.PersistentEntityResourceAssembler.toResource(PersistentEntityResourceAssembler.java:83) ~[spring-data-rest-webmvc-2.4.1.RELEASE.jar:na]
at org.springframework.data.rest.webmvc.AbstractRepositoryRestController.entitiesToResources(AbstractRepositoryRestController.java:143) ~[spring-data-rest-webmvc-2.4.1.RELEASE.jar:na]
at org.springframework.data.rest.webmvc.AbstractRepositoryRestController.toResources(AbstractRepositoryRestController.java:90) ~[spring-data-rest-webmvc-2.4.1.RELEASE.jar:na]
at org.springframework.data.rest.webmvc.AbstractRepositoryRestController.toResource(AbstractRepositoryRestController.java:110) ~[spring-data-rest-webmvc-2.4.1.RELEASE.jar:na]
at org.springframework.data.rest.webmvc.RepositorySearchController.executeSearch(RepositorySearchController.java:183) ~[spring-data-rest-webmvc-2.4.1.RELEASE.jar:na]
at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_66]
at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62) ~[na:1.8.0_66]
at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43) ~[na:1.8.0_66]
at java.lang.reflect.Method.invoke(Method.java:497) ~[na:1.8.0_66]
```

I double checked my project to ensure the spring-data-gemfire-1.8.0.BUILD-SNAPSHOT.jar file is being referenced. If I run the project with jxblum's Application.java in place, it works, if I swap it with the client factory beans, it fails (see attached file earlier).

Also I ran a gfsh command "configure pdx --read-serialized=true --disk-store" before starting up my remote GemFire server. Is it necessary?

Pleas advise.

**Comment by odrotbohm** [comment]
Hey all, I just checked the fixes and discovered that they included a few things that are worth overhauling. Generally speaking the factory beans must not register bean definitions in the application context. I'd much rather like to move to a shared `GemfireMappingContext` instance by default that can (theoretically) be customized.

I've filed [SGF-455](https://jira.spring.io/browse/SGF-455), about to be fixed in a bit. I'll ping you guys once we have builds for that fix and it would be cool if you could try this ASAP. We have a service release coming up for Gosling tomorrow about to be included into Boot 1.3.1 and it would be nice if we could ship this fix with it.

**Comment by Madder** [comment]
Switching to spring-data-gemfire-1.8.0.BUILD-SNAPSHOT did work last week but it stopped working today for some reason.
I'm now getting :
org.springframework.context.ApplicationContextException: Unable to start embedded container; nested exception is org.springframework.boot.context.embedded.EmbeddedServletContainerException: Unable to start embedded Tomcat

I don't know if it's just me...
Update: This is still working under spring-data-gemfire-1.7.2.BUILD-SNAPSHOT :)

**Comment by odrotbohm** [comment]
There should be a fixed snapshot build available by now. There was a small glitch in the fix I pushed breaking the repository detection.

**Comment by jxblum** [comment]
This should be resolved/fixed now; Closing.

**Comment by kopax** [comment]
By any chance, would you guys know why I have the same error [here](https://github.com/jclagache/spring-data-mybatis/issues/7) ?

**Comment by gregturn** [comment]
@kopax Your linked issue involves MyBatis while this guide is related to Gemfire. Instead of tacking unrelated links to an already closed ticket, please open a new one and state the details of your issue.

**Comment by kopax** [comment]
@gregturn I know it is related to mybatis, that's probably why I am not opening a new issue here. 
I have get my debbugger into the same code as you did. I have certainly less experience and I try hard to get through this task, I thought it was OK to ask here (if I am lucky) since I can't find answer on stackoverflow as you already suggested to me on another post.
